### ISSUE TITLE
Command given in tutorial "bpftrace -v -e 'tracepoint:syscalls:sys_enter_openat { printf("%s %s\n", comm, str(args->filename)); }'" fails

### ISSUE THREAD

**Issue by lyveng** [iovisor/bpftrace#855]
**Environment**:
OS: Debian 9
Kernel: 4.9.110-3+deb9u6
Since debian 9 did not have bpfcc-tools package, I downloaded the following packages from debian 10 and installed them on debian 9:
1. bpfcc-tools_0.8.0-4_all.deb
2. bpftrace_0.8+git60-gccac69c2239b-2_amd64.deb
3. libbpfcc_0.8.0-4_amd64.deb
4. libclang1-7_1%3a7.0.1-8_amd64.deb
5. libllvm7_1%3a7.0.1-8_amd64.deb
6. libtinfo6_6.1+20181013-2_amd64.deb
7. python-bpfcc_0.8.0-4_all.deb

I was following the lesson 3 from [bpftrace oneliners tutorial|https://github.com/iovisor/bpftrace/blob/master/docs/tutorial_one_liners.md]. Getting the following error
```bash
root@fk-cloud-g1a-ms-1312745:/home/livingstone.se# bpftrace -ve 'tracepoint:syscalls:sys_enter_openat { printf("%s %s\n", comm, str(args->filename)); }'
Attaching 1 probe...

Error log: 
0: (bf) r6 = r1
1: (b7) r7 = 0
2: (7b) *(u64 *)(r10 -96) = r7
3: (7b) *(u64 *)(r10 -104) = r7
4: (7b) *(u64 *)(r10 -112) = r7
5: (7b) *(u64 *)(r10 -120) = r7
6: (7b) *(u64 *)(r10 -128) = r7
7: (7b) *(u64 *)(r10 -136) = r7
8: (7b) *(u64 *)(r10 -144) = r7
9: (7b) *(u64 *)(r10 -152) = r7
10: (7b) *(u64 *)(r10 -160) = r7
11: (7b) *(u64 *)(r10 -168) = r7
12: (7b) *(u64 *)(r10 -176) = r7
13: (7b) *(u64 *)(r10 -80) = r7
14: (7b) *(u64 *)(r10 -88) = r7
15: (bf) r1 = r10
16: (07) r1 += -88
17: (b7) r2 = 16
18: (85) call 16
19: (79) r1 = *(u64 *)(r10 -88)
20: (7b) *(u64 *)(r10 -168) = r1
21: (79) r1 = *(u64 *)(r10 -80)
22: (7b) *(u64 *)(r10 -160) = r1
23: (7b) *(u64 *)(r10 -16) = r7
24: (7b) *(u64 *)(r10 -24) = r7
25: (7b) *(u64 *)(r10 -32) = r7
26: (7b) *(u64 *)(r10 -40) = r7
27: (7b) *(u64 *)(r10 -48) = r7
28: (7b) *(u64 *)(r10 -56) = r7
29: (7b) *(u64 *)(r10 -64) = r7
30: (7b) *(u64 *)(r10 -72) = r7
31: (bf) r3 = r6
32: (07) r3 += 24
33: (bf) r1 = r10
34: (07) r1 += -8
35: (b7) r2 = 8
36: (85) call 4
37: (79) r3 = *(u64 *)(r10 -8)
38: (bf) r1 = r10
39: (07) r1 += -72
40: (b7) r2 = 64
41: (85) call 45
invalid func 45

Error loading program: tracepoint:syscalls:sys_enter_openat
```

Got the following error when I ran the command from lesson 5
```bash
 bpftrace -e 'tracepoint:syscalls:sys_exit_read /pid == 18644/ { @bytes = hist(args->ret); }'
Attaching 1 probe...
ioctl(PERF_EVENT_IOC_SET_BPF): Invalid argument
Error attaching probe: tracepoint:syscalls:sys_exit_read
```

I referred a couple of other tickets similar to this like #378 and #207 
Here are the relevant details requested in those tickets

1. **debugfs**
```bash
# mount | grep sys/kernel/debug
debugfs on /sys/kernel/debug type debugfs (rw,relatime)
tracefs on /sys/kernel/debug/tracing type tracefs (rw,relatime)
# cat /sys/kernel/debug/tracing/events/syscalls/sys_enter_openat/format 
name: sys_enter_openat
ID: 540
format:
	field:unsigned short common_type;	offset:0;	size:2;	signed:0;
	field:unsigned char common_flags;	offset:2;	size:1;	signed:0;
	field:unsigned char common_preempt_count;	offset:3;	size:1;	signed:0;
	field:int common_pid;	offset:4;	size:4;	signed:1;

	field:int __syscall_nr;	offset:8;	size:4;	signed:1;
	field:int dfd;	offset:16;	size:8;	signed:0;
	field:const char * filename;	offset:24;	size:8;	signed:0;
	field:int flags;	offset:32;	size:8;	signed:0;
	field:umode_t mode;	offset:40;	size:8;	signed:0;

print fmt: "dfd: 0x%08lx, filename: 0x%08lx, flags: 0x%08lx, mode: 0x%08lx", ((unsigned long)(REC->dfd)), ((unsigned long)(REC->filename)), ((unsigned long)(REC->flags)), ((unsigned long)(REC->mode))
```
2. **Kernel Parameters**
```bash
root@fk-cloud-g1a-ms-1312745:/home/livingstone.se# grep BPF /boot/config-4.9.0-8-amd64 
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
# CONFIG_BPF_JIT_ALWAYS_ON is not set
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_NET_CLS_BPF=m
CONFIG_NET_ACT_BPF=m
CONFIG_BPF_JIT=y
CONFIG_HAVE_EBPF_JIT=y
CONFIG_BPF_EVENTS=y
CONFIG_TEST_BPF=m
```
3. **perf output**
```bash
# perf stat -e syscalls:sys_exit_open* -a -- sleep 10

 Performance counter stats for 'system wide':

                 0      syscalls:sys_exit_open_by_handle_at                                   
             28674      syscalls:sys_exit_open                                      
                 0      syscalls:sys_exit_openat                                    

      10.014810684 seconds time elapsed

```
4. **strace output**
I'm not pasting the entire output. Just pasting lines that I though might be relevant
```bash
# strace bpftrace -e 'tracepoint:syscalls:sys_enter_openat { printf("%s %s\n", comm, str(args->filename)); }'
bpf(BPF_MAP_UPDATE_ELEM, {map_fd=3, key=0x7ffc3807ae7c, value=0x7ffc3807ae80, flags=BPF_ANY}, 112) = 0
epoll_ctl(4, EPOLL_CTL_ADD, 52, {EPOLLIN, {u32=1646178688, u64=94314833031552}}) = 0
dup(2)                                  = 53
open("/dev/null", O_WRONLY)             = 54
dup2(54, 2)                             = 2
close(54)                               = 0
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=264558}, 112) = -1 E2BIG (Argument list too long)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=264558}, 112) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff75c6d4f0, kern_version=264558}, 112) = -1 EINVAL (Invalid argument)
write(2, "bpf: Failed to load program: Inv"..., 46) = 46
write(2, "0: (bf) r6 = r1\n1: (b7) r7 = 0\n2"..., 1128) = 1128
write(2, "HINT: bpf_probe_read_str missing"..., 57) = 57
uname({sysname="Linux", nodename="fk-cloud-g1a-ms-1312745", ...}) = 0
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=264448}, 112) = -1 E2BIG (Argument list too long)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=264448}, 112) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff75c6d4f0, kern_version=264448}, 112) = -1 EINVAL (Invalid argument)
write(2, "bpf: Failed to load program: Inv"..., 46) = 46
write(2, "0: (bf) r6 = r1\n1: (b7) r7 = 0\n2"..., 1128) = 1128
write(2, "HINT: bpf_probe_read_str missing"..., 57) = 57
open("/usr/include/linux/version.h", O_RDONLY) = 54
read(54, "#define LINUX_VERSION_CODE 26459"..., 8191) = 97
read(54, "", 8191)                      = 0
close(54)                               = 0
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=264592}, 112) = -1 E2BIG (Argument list too long)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=264592}, 112) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=70, insns=0x7f1438e1e000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff75c6d4f0, kern_version=264592}, 112) = -1 EINVAL (Invalid argument)
write(2, "bpf: Failed to load program: Inv"..., 46) = 46
write(2, "0: (bf) r6 = r1\n1: (b7) r7 = 0\n2"..., 1128) = 1128
write(2, "HINT: bpf_probe_read_str missing"..., 57) = 57
dup2(53, 2)                             = 2
close(53)                               = 0
futex(0x7f142fdd41a0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
write(2, "Error loading program: tracepoin"..., 68) = 68
write(2, "\n", 1)                       = 1
munmap(0x7f1438e1e000, 4096)            = 0
munmap(0x7f1438e1d000, 4096)            = 0
brk(0x5585e6a4d000)                     = 0x5585e6a4d000
close(3)                                = 0
exit_group(-1)                          = ?
+++ exited with 255 +++
```

**Comment by brendangregg** [comment]
At first glance it looks like a kernel headers mismatch (the "call 45" being invalid). The strace output has "HINT: bpf_probe_read_str missing" (why wasn't that in the normal output?) so it sounds like bpf_probe_read_str() is call 45, and the kernel doesn't have it, but the header files bpftrace has been built for has. Like everything in BPF, these were added in parts. bpf_probe_read_str() was added in 4.11 (https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md ).

AFAIK, the main development here is on Ubuntu, Fedora, RHEL, and ArchLinux, so you might be the only person with a handy environment to debug this further.

**Comment by lyveng** [comment]
@brendangregg I've updated the strace output to include a couple of more relevant lines. Here is what I could figure out from the strace output
1. dup(2) returned fd 53
2. open("/dev/null") returned fd 54
3. dup2(54, 2) returned fd 2 effectively redirecting stderr to /dev/null
4. close(54)
So now fd 2, i.e., stderr points to /dev/null and fd 54 points to actual stderr
So when the HINT was written, it was written to fd 2 which was effectively pointing to /dev/null
Just after that 
5. dup2(53, 2) returns fd 2. This restored the actual stderr to fd 2.
Whatever we see in the output is printed after this. This is why I think the HINT is missing in the output. Though I have no idea why bpftrace is written like that.

Regarding the actual problem I think what you suggested is correct. As I have mentioned in the environment section, I had downloaded the bpftrace package that was built for debian 10, i.e., 4.19 kernel and installed it on debian 9, i.e., 4.9 kernel with the corresponding header files installed since debian 9 did not have the bpf packages. I think that is the reason it is not working. I'm planning to upgrade the kernel to the one from stretch-backports which is 4.19+105~bpo9+1 and install the corresponding headers. Maybe that will get it working.
Thanks for the help. Will check and get back.

But I think it would be better off to have this mentioned([BPF Features by Linux Kernel Version](https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md#bpf-features-by-linux-kernel-version)) in the one liners tutorial so that new bpf newbies don't get confused.
